### DISCUSSION TITLE
gatsby-transformer-remark - unable to filter `allMarkdownRemark` nodes by parent `File` node `sourceInstanceName`.

### DISCUSSION THREAD

**Discussion by kimbaudi** [gatsbyjs/gatsby#30395] [Help] [answered]
## Description

`MarkdownRemark` nodes already provide `fileAbsolutePath`, but it would be really useful to also provide parent File node `sourceInstanceName`, `relativeDirectory`, `name` in `MarkdownRemark` nodes as `fileSourceInstanceName`, `fileRelativeDirectory`, and `fileName`.

Gatsby provides `gatsby-source-filesystem` and `gatsby-transformer-remark` plugins to source markdown files and generate html pages.

**gatsby-config.js**

```js
module.exports = {
  plugins: [
    {
      resolve: `gatsby-source-filesystem`,
      options: {
        name: `blog-posts`,
        path: path.resolve(__dirname, `./src/contents/blog-posts`),
      },
    },
    {
      resolve: `gatsby-transformer-remark`,
      options: {
        excerpt_separator: `<!--more-->`,
      },
    },
  ],
}
```

Right now, I am filtering `MarkdownRemark` nodes using glob on `fileAbsolutePath` field to get all markdowns under `src/contents/blog-posts` folder. However, I want to filter `MarkdownRemark` nodes using parent `File` node `sourceInstanceName` field.

![a](https://user-images.githubusercontent.com/1864092/112020992-12f7a080-8aee-11eb-96f8-a737e89c489f.png)

```gql
{
  allBlogPost: allMarkdownRemark(
    filter: {fileAbsolutePath: {glob: "**/blog-posts/**"}}
    sort: {fields: frontmatter___date, order: DESC}
  ) {
    blogPosts: nodes {
      fileAbsolutePath
      fields {
        slug
      }
      parent {
        ... on File {
          sourceInstanceName
        }
      }
      frontmatter {
        title
        author
        date(formatString: "MMMM DD, YYYY")
        tags
      }
    }
  }
}
```

Unfortunately, I cannot filter on parent other than id so I am unable to filter `MarkdownRemark` nodes by `sourceInstanceName`.

![b](https://user-images.githubusercontent.com/1864092/112021023-18ed8180-8aee-11eb-97ce-4091b2b69a03.png)

```gql
{
  allBlogPost: allMarkdownRemark(
    filter: {
      parent: {sourceInstanceName: "blog-posts"}
    }
    sort: {fields: frontmatter___date, order: DESC}
  ) {
    blogPosts: nodes {
      fileAbsolutePath
      fields {
        slug
      }
      parent {
        ... on File {
          sourceInstanceName
        }
      }
      frontmatter {
        title
        author
        date(formatString: "MMMM DD, YYYY")
        tags
      }
    }
  }
}
```

It would be great if `MarkdownRemark` provided `fileSourceInstanceName` as well as `fileRelativeDirectory` and `fileName` so we can filter `allMarkdownRemark` using those fields:

```gql
{
  allBlogPost: allMarkdownRemark(
    filter: {fileSourceInstanceName: {eq: "blog-posts"}}
    sort: {fields: frontmatter___date, order: DESC}
  ) {
    blogPosts: nodes {
      fileAbsolutePath
      fileSourceInstanceName
      fileRelativeDirectory
      fileName
    }
  }
}
```

or provide a file node with those fields:

```gql
{
  allBlogPost: allMarkdownRemark(
    filter: {file: sourceInstanceName: {eq: "blog-posts"}}
    sort: {fields: frontmatter___date, order: DESC}
  ) {
    blogPosts: nodes {
      file {
        absolutePath
        sourceInstanceName
        relativeDirectory
        name
      }
    }
  }
}
```

I think having the ability to filter `allMarkdownRemark` nodes by File `sourceInstanceName` would be a cleaner alternative to globbing `fileAbsolutePath`.

### Steps to reproduce

I've made a minimal repo at https://github.com/kimbaudi/gatsby-group-query. 

### Expected result

Have ability to filter `allMarkdownRemark` nodes by file `sourceInstanceName` field

### Actual result

Only have ability to filter `allMarkdownRemark` nodes by `fileAbsolutePath` field and not `sourceInstanceName` field (or file `relativeDirectory`, `name` fields)

```gql
{
  allBlogPost: allMarkdownRemark(
    filter: { fileAbsolutePath: { glob: "**/blog-posts/**" } }
  ) {
    blogPosts: nodes {
      fileAbsolutePath
    }
  }
}
```

**Comment by kimbaudi** [comment]
linking to discussion https://github.com/gatsbyjs/gatsby/discussions/30393

**Comment by LekoArts** [comment, marked as answer]
You can use https://www.gatsbyjs.com/docs/reference/config-files/actions/#createNodeField to add additional fields

**Comment by kimbaudi** [comment]
@LekoArts - thank you for your suggestion... it works perfectly!

For anyone else interested, here is my `onCreateNode` using `createNodeField` to add `absolutePath`, `name` and `sourceInstanceName` fields:

**gatsby-node.js**

```js
exports.onCreateNode = ({ node, actions: { createNodeField }, getNode }) => {
  if (node.internal.type === `MarkdownRemark`) {
    const { absolutePath, name, sourceInstanceName } = getNode(node.parent)
    const relativeFilePath = createFilePath({
      node,
      getNode,
    })
    createNodeField({
      name: `absolutePath`,
      node,
      value: absolutePath,
    })
    createNodeField({
      name: `name`,
      node,
      value: name,
    })
    createNodeField({
      name: `sourceInstanceName`,
      node,
      value: sourceInstanceName,
    })
    createNodeField({
      name: `slug`,
      node,
      value: `/${sourceInstanceName}${relativeFilePath}`,
    })
  }
}
```

And now I can filter those fields:

![awesome](https://user-images.githubusercontent.com/1864092/112033315-87384100-8afa-11eb-801a-b92616d0dbd2.png)
